provider/pkg/metadata/naming.go

```diff
@@ -15,22 +15,36 @@
 package metadata
 
 import (
+	"errors"
+
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // AssignNameIfAutonamable generates a name for an object. Uses DNS-1123-compliant characters.
 // All auto-named resources get the annotation `pulumi.com/autonamed` for tooling purposes.
-func AssignNameIfAutonamable(randomSeed []byte, obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) {
+func AssignNameIfAutonamable(randomSeed []byte, engineAutonaming *pulumirpc.CheckRequest_AutonamingOptions,
+	obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) error {
 	contract.Assertf(urn.Name() != "", "expected non-empty name in URN: %s", urn)
-	if !IsNamed(obj, propMap) && !IsGenerateName(obj, propMap) {
-		prefix := urn.Name() + "-"
-		autoname, err := resource.NewUniqueName(randomSeed, prefix, 0, 0, nil)
-		contract.AssertNoErrorf(err, "unexpected error while creating NewUniqueName")
-		obj.SetName(autoname)
-		SetAnnotationTrue(obj, AnnotationAutonamed)
+	if IsNamed(obj, propMap) || IsGenerateName(obj, propMap) {
+		return nil
+	}
+	prefix := urn.Name() + "-"
+	autoname, err := resource.NewUniqueName(randomSeed, prefix, 0, 0, nil)
+	contract.AssertNoErrorf(err, "unexpected error while creating NewUniqueName")
+	if engineAutonaming != nil {
+		switch engineAutonaming.Mode {
+		case pulumirpc.CheckRequest_AutonamingOptions_DISABLE:
+			return errors.New("autonaming is disabled, please provide a name or generateName value")
+		case pulumirpc.CheckRequest_AutonamingOptions_ENFORCE, pulumirpc.CheckRequest_AutonamingOptions_PROPOSE:
+			autoname = engineAutonaming.ProposedName
+		}
```

Comment:
Is it definitely true that ProposedName is non-empty in these modes?  The logic isn't totally clear about which modes would use a provider-generated value, would be nice to be more explicit. 
In other words, is it possible that the mode would be "proposed" without an actual proposed value?